apps/call.go

```diff
@@ -47,163 +28,65 @@ type Call struct {
-// triggers displaying a Modal. Returning a "call" type in response to a
-// submission causes the call to be executed from the user-agent (NOT
-// IMPLEMENTED YET)
-//
-// Form requests expect form or error.
-//
-// Lookup requests expect ok or error.
-//
-// In case of an error, the returned response type is "error", ErrorText
-// contains the overall error text. Data contains optional, field-level errors.
-type CallResponse struct {
-	Type CallResponseType `json:"type"`
-
-	// Used in CallResponseTypeOK to return the displayble, and JSON results
-	Markdown string      `json:"markdown,omitempty"`
-	Data     interface{} `json:"data,omitempty"`
-
-	// Used in CallResponseTypeError
-	ErrorText string `json:"error,omitempty"`
-
-	// Used in CallResponseTypeNavigate
-	NavigateToURL      string `json:"navigate_to_url,omitempty"`
-	UseExternalBrowser bool   `json:"use_external_browser,omitempty"`
-
-	// Used in CallResponseTypeCall
-	Call *Call `json:"call,omitempty"`
-
-	// Used in CallResponseTypeForm
-	Form *Form `json:"form,omitempty"`
-}
-
-func NewErrorResponse(err error) CallResponse {
-	return CallResponse{
-		Type: CallResponseTypeError,
-		// TODO <>/<> ticket use MD instead of ErrorText
-		ErrorText: err.Error(),
-	}
-}
-
-func NewDataResponse(data interface{}) CallResponse {
-	return CallResponse{
-		Type: CallResponseTypeOK,
-		Data: data,
+func (c *Call) UnmarshalJSON(data []byte) error {
+	stringValue := ""
+	err := json.Unmarshal(data, &stringValue)
+	if err == nil {
+		*c = Call{
+			Path: stringValue,
+		}
+		return nil
 	}
-}
 
-func NewTextResponse(format string, args ...interface{}) CallResponse {
-	return CallResponse{
-		Type:     CallResponseTypeOK,
-		Markdown: fmt.Sprintf(format, args...),
+	// Need a type that is just like Call, but without UnmarshalJSON
+	structValue := struct {
+		Path   string      `json:"path,omitempty"`
+		Expand *Expand     `json:"expand,omitempty"`
+		State  interface{} `json:"state,omitempty"`
+	}{}
+	err = json.Unmarshal(data, &structValue)
+	if err != nil {
+		return err
 	}
-}
 
-func NewFormResponse(form Form) CallResponse {
-	return CallResponse{
-		Type: CallResponseTypeForm,
-		Form: &form,
+	*c = Call{
+		Path:   structValue.Path,
+		Expand: structValue.Expand,
+		State:  structValue.State,
 	}
+	return nil
 }
 
-func NewLookupResponse(opts []SelectOption) CallResponse {
-	return NewDataResponse(struct {
-		Items []SelectOption `json:"items"`
-	}{opts})
-}
-
-// Error() makes CallResponse a valid error, for convenience
-func (cr CallResponse) Error() string {
-	if cr.Type == CallResponseTypeError {
-		return cr.ErrorText
+func NewCall(path string) *Call {
+	c := Call{
+		Path: path,
 	}
-	return ""
+	return &c
 }
 
-func CallRequestFromJSON(data []byte) (*CallRequest, error) {
-	c := CallRequest{}
-	err := json.Unmarshal(data, &c)
-	if err != nil {
-		return nil, err
-	}
-	return &c, nil
+func (c Call) WithExpand(expand Expand) *Call {
+	c.Expand = &expand
+	return &c
 }
```

Comment:
Can we simplify this to 
```suggestion
func (c Call) WithExpand(expand Expand) Call {
	c.Expand = &expand
	return c
}
```
?

Reply:
returning a *Call was intentional since a lot of the internal functions and structs take a *Call, and you can't do `&NewCall()`